Add OpenMPI variants to MOOSE development packages, add Python 3.11 support

milljm opened the pull request:
Add OpenMPI variants to our Conda packages.

This will have the side effect of marching towards Python 3.11.


Closes #26839

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2069898/) on bdd72d9 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2070894/) on 4c2162f wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2072044/) on 09df209 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2072776/) on 01971f6 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2079126/) on f8273dc wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2079682/) on 7f5c408 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2093326/) on e405de8 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2093392/) on 92f2144 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2094989/) on 570a8a8 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2096348/) on aa511f5 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2100239/) on d36ae0c wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2103543/) on 46ade1c wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2105082/) on 56d192a wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2119155/) on 6eb7d17 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2123350/) on 5698d94 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2125334/) on a93f83e wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2125334/) on a93f83e wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2125429/) on d6ab931 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2125523/) on 8be2fa6 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2132815/) on f76f98b wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2164707/) on fcdb79e wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2164763/) on 86566f4 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2164857/) on 3997dd4 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Add Conda OpenMPI variant, Python 3.11 support

Bump MPICH 4.1.2 --> 4.2.1
Bump Clang 12 --> 16

Allow petsc configure/build to retry on specific conda related
errors.

Add ability to "try again" during a conda build. I want to do this
here rather than on Civet recipes because I was running into
failures on my work machine due to libraries not being loaded. I
think this is a Conda bug/issue, as it also happens on my personal
machine.

Modify premake.py to make use of new variant naming conventions.

Update diagnostics.sh script to use new Conda naming conventions.

Closes #26839

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2182655/) on 3c40b4c wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2182768/) on 09253b6 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2183460/) on 9153b70 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2183884/) on c782f8b wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2184329/) on d874fd1 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2185076/) on 444a8d4 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2191597/) on 737e6c4 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2192253/) on 23f0540 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2193239/) on 97d311d wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2194896/) on 090ff3d wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2194892/) on 090ff3d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26848/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2194936/) on 090ff3d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>53c560</th>
      <th colspan=4>#26848 090ff3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.13%</td>
      <td>85.13%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>103068</td>
      <td>103067</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18004</td>
      <td>18005</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26848/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26848/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26848/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2200685/) on 34b1fe2 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2204179/) on 8953c71 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

milljm commented on conda/libmesh-vtk/build.sh:
```diff
@@ -18,11 +18,13 @@ cmake .. -G "Ninja" \
     -DVTK_GROUP_ENABLE_Rendering:STRING=DONT_WANT \
     -DVTK_GROUP_ENABLE_Qt::STRING=NO \
     -DVTK_GROUP_ENABLE_Views:STRING=NO \
-    -DVTK_GROUP_ENABLE_Web:STRING=NO 
```
@cticenhour this is what I was asking you for when I mentioned I was getting GLIBC issues! This _is the fix_.

The error:
```pre
nstalled/lib/libmesh_opt.so, not found (try using -rpath or -rpath-link)
Compiling C++ (in opt mode) /home/milljm/projects/moose/framework/build/unity_src/samplers_Unity.C...
/home/milljm/miniforge_testing/envs/testing/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/milljm/miniforge_testing/envs/testing/lib/libvtksys-9.2.so: undefined reference to `clock_gettime@GLIBC_2.17'
/home/milljm/miniforge_testing/envs/testing/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/milljm/miniforge_testing/envs/testing/lib/libvtkIOCore-9.2.so: undefined reference to `memcpy@GLIBC_2.14'
```

Does this make any sense that these switches would solve this issue!?

milljm commented on conda/libmesh-vtk/build.sh:
```diff
@@ -18,11 +18,13 @@ cmake .. -G "Ninja" \
+    -DVTK_GROUP_ENABLE_Web:STRING=NO \
+    -DVTK_GROUP_ENABLE_MPI=NO \
+    -DVTK_USE_MPI:BOOL=OFF \
+    -DVTK_MODULE_ENABLE_VTK_IOMPIParallel=WANT
```
@cticenhour this is what I was asking you for when I mentioned I was getting GLIBC issues! This _is the fix_.

The error:
```pre
nstalled/lib/libmesh_opt.so, not found (try using -rpath or -rpath-link)
Compiling C++ (in opt mode) /home/milljm/projects/moose/framework/build/unity_src/samplers_Unity.C...
/home/milljm/miniforge_testing/envs/testing/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/milljm/miniforge_testing/envs/testing/lib/libvtksys-9.2.so: undefined reference to `clock_gettime@GLIBC_2.17'
/home/milljm/miniforge_testing/envs/testing/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/milljm/miniforge_testing/envs/testing/lib/libvtkIOCore-9.2.so: undefined reference to `memcpy@GLIBC_2.14'
```

Does this make any sense that these switches would solve this issue!?

cticenhour commented on scripts/versioner.py:
```diff
@@ -46,24 +46,9 @@ def undefined(arg, *args, **kwargs):
-                'base_mpicc'            : undefined('mpicc'),
-                'base_mpicxx'           : undefined('mpicxx'),
-                'base_mpifort'          : undefined('mpifort'),
+                'mpi'                   : undefined('mpich|openmpi'),
```
Only one question on this- no need to check `mpicc, mpicxx, mpifort` etc. here since we're checking against the variants? 

Otherwise nice simplification in `moose-mpi` and here!

milljm replied:
This is for both assigning a value to variables encountered by jinja templating, and for cosmetic reasons. The `mpi` key/value is actually not valid of course, but it allows us to print out something like this:

```bash
$ cd moose/scripts/
$ ./versioner.py mpi --yaml | grep install
  install: moose-mpi=2024.04.25=mpich|openmpi
$ 
```
I realize some people are going to literally copy/paste that entire line... But I am not sure how else to inform a person that they have a choice in the matter. I am totally open to suggestions!

milljm replied:
Spaces work here:
```bash
$ ./versioner.py mpi --yaml | grep install
  install: moose-mpi=2024.04.25=mpich or openmpi
```

milljm replied:
Oh, I have a better idea. I can read our `packages_config.yml` file to determine our "default" MPI wrapper of choice.

cticenhour replied:
Thanks for the clarification of what that bit of code is actually doing. Reading from `packages_config.yml` is fine to indicate a default I think. Though it would be nice to show that openmpi is an option for those who go far enough to run that command. 

Perhaps something like:

```
$ ./versioner.py mpi --yaml | grep install
  install: moose-mpi=2024.04.25={mpich || openmpi}
```

This would indicate to a developer the `OR`-ness of this choice to me in logic that we're used to seeing.

cticenhour commented on scripts/versioner.yaml:
```diff
@@ -4,18 +4,30 @@
+      - conda/mpi/conda_build_config.yaml
+      - conda/mpi/build.sh
+    apptainer:
+  # Deprecated. The following key (mpich) is only necessary for Civet recipes until the PR
+  # responsible for this change makes it all the way into master.
+  # TODO: remove mpich when able. When you do, remove the same deprecations in test_versioner.py
+  #       (line 49) and versioner.py (101)
```
👍🏻

cticenhour commented on conda/petsc/build.sh:
```diff
@@ -1,119 +1,164 @@
-set -eu
-export PATH=/bin:$PATH
-
+#
```
This is some awesome culling of unnecessary stuff! That being said, we should remove the comments.

cticenhour commented on conda/petsc/build.sh:
```diff
@@ -1,119 +1,164 @@
+#    FCFLAGS="${FCFLAGS} -I$PREFIX/include"
+#fi
+#
+## Remove std=C++17 from CXXFLAGS as we specify the C++ dialect for PETSc as C++17 in configure_petsc.
+## Specifying both causes an error as of PETSc 3.17.
```
```suggestion
# Remove std=C++17 from CXXFLAGS as we specify the C++ dialect for PETSc as C++17 in configure_petsc.
# Specifying both causes an error as of PETSc 3.17.
```

cticenhour commented on conda/petsc/build.sh:
```diff
@@ -1,119 +1,164 @@
+#fi
+#
+## Handle switches created by Conda variants
+ADDITIONAL_ARGS=""
```
Where is this getting defined? You've commented out line 63. Is this (and its presence in the PETSC config line below) still necessary?

milljm replied:
I think I need to leave it, as I believe it is not possible to comment out a line continuation later:

```bash
configure_petsc \
      --COPTFLAGS=-O3 \
      --CXXOPTFLAGS=-O3 \
      --FOPTFLAGS=-O3 \
      --with-x=0 \
      --with-ssl=0 \
      --with-hdf5-dir=$HDF5_DIR \
      --with-mpi-dir=$PREFIX \
      AR="$AR" \
      RANLIB="$RANLIB" \
      CFLAGS="$CFLAGS" \
      CXXFLAGS="$CXXFLAGS" \
      CPPFLAGS="$CPPFLAGS" \
      FFLAGS="$FFLAGS" \
      FCFLAGS="$FFLAGS" \
      LDFLAGS="$LDFLAGS" \
      ${ADDITIONAL_ARGS} \         <----------------------------------- this guy
      --prefix=$PREFIX/petsc || (tail -400 configure.log && exit 1)
```
I could just remove both until the time is needed (this will be needed for pytorch stuff soon (tm).

cticenhour replied:
It's fine to leave it in if you leave a good TODO comment about why it's there and empty.

cticenhour commented on conda/petsc/build.sh:
```diff
@@ -1,119 +1,164 @@
+#if [[ $mpi == "mpich" ]] && [[ $(uname) == Linux ]]; then
+#  make check
+#fi
+make PETSC_DIR=$SRC_DIR PETSC_ARCH=$PETSC_ARCH all
+make PETSC_DIR=$SRC_DIR PETSC_ARCH=$PETSC_ARCH install
```
`make check` bites the dust ☠️ 

Nice touch in robustify-ing the `make` lines with the env variables.

milljm replied:
I'll put it back :) Maybe with the whole
```bash
FI_PROVIDER=tcp
```
it will now work!

cticenhour commented on conda/petsc/build.sh:
```diff
@@ -1,119 +1,164 @@
+#rm -f ${PREFIX}/moose-petsc/lib/petsc/conf/configure-hash
+#find ${PREFIX}/moose-petsc/lib/petsc -name '*.pyc' -delete
+#
+## Replace ${BUILD_PREFIX} after installation,
+## otherwise 'make install' above may fail
```
```suggestion
# Replace ${BUILD_PREFIX} after installation,
# otherwise 'make install' above may fail
```

cticenhour commented on conda/petsc/build.sh:
```diff
@@ -1,119 +1,164 @@
-du -hs $PREFIX/share/petsc/datafiles/*
-rm -fr $PREFIX/share/petsc/datafiles
-
+#
+#echo "Removing example files"
+#du -hs $PREFIX/moose-petsc/share/petsc/examples/src
+#rm -fr $PREFIX/moose-petsc/share/petsc/examples/src
+#echo "Removing data files"
+#du -hs $PREFIX/moose-petsc/share/petsc/datafiles/*
+#rm -fr $PREFIX/moose-petsc/share/petsc/datafiles
+#
```
No longer removing examples and datafiles?

milljm replied:
Hmmm... sounds like we should do this. I'll put these things back!

cticenhour reviewed (commented):
Making this a comment rather than finalizing the review so we can chat over this instead of me locking in "requests changes". Congrats on getting this working altogether! Specific comments below, but all of these are questions or clean-up needs.

milljm commented on conda/template/meta.yaml:
```diff
@@ -23,9 +23,11 @@ build:
 
 requirements:
   build:
+    - openmpi
```
Good call. Originally we were going to need to switch to OpenMPI because we didn't know why latest versions of MPICH was hanging.

That was until @lindsayad discovered the following solves that issue:
```bash
 FI_PROVIDER=tcp
```
*which is now being set in our `moose-mpi` package.*

I agree, we should switch this back to mpich!

milljm commented on framework/doc/packages_config.yml:
```diff
@@ -10,8 +10,8 @@ maximum_python: 3.11
+moose_dev: 2024.04.25
+moose_tools: 2024.04.25
 apptainer_rockylinux: 8.8.20230518
 apptainer_mpich: 3.4.3
```
Confirmed...
```bash
$ > moosetainer shell moose-dev

[moose-dev][/opt/mpich/bin]$ ./mpichversion 
MPICH Version:    	3.4.3
```
Changing this will have radical changes in our apptainer containers. Very much equivalent to what this PR is doing in our Conda packages. Unless we have a real good reason for bumping it, I suggest leaving well enough alone...

cticenhour replied:
Yeah, we can tackle that in a separate PR down the road.

milljm commented on conda/mpi/meta.yaml:
```diff
@@ -0,0 +1,83 @@
+about:
+  home: https://mooseframework.org/
+  license: LGPL 2.1
+  summary: 'A top-level package designed to control environment variables related to conda-forge provided MPICH'
+  description: |
+    MPICH is a high performance and widely portable implementation of the
+    Message Passing Interface (MPI) standard.
+  doc_url: http://www.mpich.org/documentation/guides
+  dev_url: https://wiki.mpich.org/mpich/index.php/Main_Page
```
Wow! I'll try. Upon first glance, I don't see why not. I should be able to use the conda_build_config.yaml file to populate this in zip keys or some such.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2207113/) on 408f87d wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2207209/) on fbd9721 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

milljm commented on framework/doc/packages_config.yml:
```diff
@@ -7,15 +7,16 @@ maximum_clang: 16.0.6
-moose_dev: 2024.04.23
-moose_tools: 2024.03.07
+moose_dev: 2024.04.25
+moose_tools: 2024.04.25
```
damn it! forgot...

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2207285/) on 8b1afdb wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2207383/) on 86b89c5 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

cticenhour commented on conda/petsc/build.sh:
```diff
@@ -56,15 +39,17 @@ configure_petsc \
     FFLAGS="$FFLAGS" \
     FCFLAGS="$FCFLAGS" \
```
Actually, looks like we used to say FCFLAGS = FFLAGS to begin with and then modify further. Would you like to set this line to `FCFLAGS="$FFLAGS"` instead of removing it?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2207718/) on e3ff207 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2207718/) on e3ff207 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2212521/) on 0ddfcb2 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2212521/) on 0ddfcb2 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2218824/) on f8eccb7 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2231663/) on 7a4bda0 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2231910/) on 9927fa3 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2231910/) on 9927fa3 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2233606/) on 32bee33 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2243788/) on d056a4d wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

lindsayad committed:
Update tests for vtk changes from 9.2.x to 9.3

Threshold tests can be fixed without regolding by adjusting the
threshold. All other tests had to be regolded. Relevant VTK MR
is https://gitlab.kitware.com/vtk/vtk/-/merge_requests/9361. The
TLDR is that the vtk MR gave us more precision

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2254491/) on 1166012 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
modify apptainer_generator.py

More necessary changes to allow apptainer_generator.py to work with
variants.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2256400/) on 5291972 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

lindsayad commented:
The Conda ARM Mac failures look concerning 😭 It looks like if we fix that scripts test we'll be down to 1, maybe 2 failing recipes depending on whether the Conda Linux target failure is reproducible

Commit:
Modify unittests to work with variants

unittests to work with variants: Not exactly. What I did was alter
checks against build_string to build_number instead. When I get a
better handle on premake.py, I will come back and factor in variants
properly.

Have Peacock create a proper build_string varaint name as our other
packages.

diagnostic_environment.sh: I found when python calls subprocess to
execute a function, that function does not exist in the subprocess's
environment, and so fails with "command not found". To work around
this issue, use the declared CONDA_EXE, as that is the proper pointer
to the binary `conda`. TODO: This may need to be adjusted in
premake.py as well.

Closes #26839

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2259078/) on a5be793 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
looks like libzlib was updated

Need to pin older libzlib since pyqt/vtk are still using 1.2.13.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2259301/) on b48bb77 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2259341/) on b48bb77 wanted to post the following:


The following coverage requirement(s) failed:
- `electromagnetics` coverage rate 93.52% is less than the required 94.0%
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- `thermal_hydraulics` coverage rate 88.36% is less than the required 89.0%

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2259341/) on b48bb77 wanted to post the following:


The following coverage requirement(s) failed:
- `electromagnetics` coverage rate 93.52% is less than the required 94.0%
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- `thermal_hydraulics` coverage rate 88.36% is less than the required 89.0%

Commit:
Skip incompatible archs

The following modified tests are not capable of running
on Clang 16.x Apple Si.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2261389/) on d0d41be wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2261389/) on d0d41be wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Additional Skips for ARM64

TODO: Lets talk about these before allowing them in.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2262861/) on 3a718ec wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2262901/) on 3a718ec wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

Commit:
Fix unittests

WIP: unit/src/MooseUtilsTest.C:129: Failure

lindsayad committed:
Do double cast check

best ref I've found is https://github.com/golang/go/issues/47387

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2264799/) on e7ea45b wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2268679/) on db5abe0 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2268892/) on 9f603c0 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2272456/) on 728a2e7 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2272765/) on 3b5b50c wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Rebase on next for moose-tools update

rebase/bump dates. Versioner hashes to follow.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2285207/) on 53cb7c3 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Pin gcc and gcc_impl_linux-64

It seems like build_10 is broken?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2287780/) on 9c37359 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2288030/) on ae85d87 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2288030/) on ae85d87 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
pin gfortran_impl as well

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2289016/) on 4ee0df4 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2294384/) on 6e9ded3 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2294410/) on 6e9ded3 : invalidated by @milljm

A timeout test test_phasefield threads failed. 4 hours

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2294424/) on 6e9ded3 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 89.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2294424/) on 6e9ded3 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 85.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 89.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

lindsayad commented on conda/libmesh-vtk/build.sh:
```diff
@@ -3,36 +3,44 @@ set -eu
-ninja install -v -j ${MOOSE_JOBS:-2}
+# Tired of failing on build events that can be fixed by an invalidation on Civet.
+# Handle retries for this one step so as to not need an entire 4 hour build target redo.
+TRY_AGAIN_REASON='Library not loaded: @rpath/'
```
what is the usual cause for the failure?

milljm replied:
The things we sometimes see during Conda build. And when actually using the packages. I am sure you've come across the following. Didn't take me long to peruse the Next build history to find one:

```bash
Compiling Fortan Plugin (in opt mode) /opt/civet0/build_0/moose/modules/solid_mechanics/test/plugins/elastic_dpredef.f...
dyld[1507]: Library not loaded: @rpath/libLLVM-12.dylib
```

Ref: https://civet.inl.gov/job/2284808/

lindsayad commented on conda/mpi/build.sh:
```diff
@@ -10,11 +10,14 @@ TEMP_CXXFLAGS=${CXXFLAGS//-std=c++[0-9][0-9]}
-export CC=mpicc CXX=mpicxx FC=mpif90 F90=mpif90 F77=mpif77 C_INCLUDE_PATH=${PREFIX}/include MOOSE_NO_CODESIGN=true MPIHOME=${PREFIX} CXXFLAGS="$ACTIVATION_CXXFLAGS" HDF5_DIR=${PREFIX}
+export CC=mpicc CXX=mpicxx FC=mpif90 F90=mpif90 F77=mpif77 C_INCLUDE_PATH=${PREFIX}/include \
+ MOOSE_NO_CODESIGN=true MPIHOME=${PREFIX} CXXFLAGS="$ACTIVATION_CXXFLAGS" HDF5_DIR=${PREFIX} \
+ FI_PROVIDER=tcp
```
this is almost surely not needed for openmpi but I don't see a lot of harm in exporting it anyway

lindsayad commented on conda/petsc/repeat_build.sh:
```diff
@@ -0,0 +1,25 @@
+source $SRC_DIR/configure_petsc.sh
+configure_petsc \
+    --COPTFLAGS=-O3 \
+    --CXXOPTFLAGS=-O3 \
+    --FOPTFLAGS=-O3 \
```
How'd we come up with these?

lindsayad replied:
`-O3` is a logical choice. I'm just commenting because we do `-O2` in libmesh/MOOSE and IIRC @roystgnr may have said we see now benefits from `-O3` or even a slow-down(?)

milljm replied:
When we borrowed sections from petsc-feedstock on conda-forge when we first created this package.

Ref: https://github.com/conda-forge/petsc-feedstock/blob/main/recipe/build.sh

grmnptr replied:
Yeah pretty cool!

Commit:
do not pin openssl / matplotlib pinnings

Fix conflicts that will ensue if we allow openssl to be pinned during
moose-mpi package creation.

Adding matplotlib pinnings. While this is something we're trying to do
in another PR (libtorch) this PR seems to have the right-of-way.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2310876/) on 9d412f9 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

lindsayad reviewed (dismissed):
assuming this passes testing

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2310962/) on c766910 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Rebas on next, bump dates

bump to versioner hases to follow

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2313237/) on bfd76b1 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2315265/) on 11d0eb6 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/2315298/) on 11d0eb6 : invalidated by @milljm

fatal: unable to access &#x27;&#x27;https://code.ornl.gov/neams/saline/&#x27;&#x27;: Could not resolve host: code.ornl.gov

lindsayad commented:
looks good to me!

loganharbour commented on conda/libmesh-vtk/repeat_build.sh:
```diff
@@ -0,0 +1,32 @@
+# Settings guide: https://docs.vtk.org/en/latest/build_instructions/build_settings.html
```
Either need to add to versioner or move back. Either is fine with me!

loganharbour commented on conda/libmesh/repeat_build.sh:
```diff
@@ -0,0 +1,8 @@
+LIBMESH_DIR=${PREFIX}/libmesh \
```
Same thing for this one (versioner or move)

loganharbour commented on conda/petsc/build.sh:
```diff
@@ -1,70 +1,66 @@
-  if grep gcc_ext $f; then
-    echo "gcc_ext found in $f"
+# Manualy download these troublesome packages (due to our network)
+mkdir downloads; cd downloads
```
We discussed that we could probably remove these

loganharbour reviewed (changes requested):
All done

Commit:
remove repeat_build script, shell-linting

Remove repeat build scripts in favor of including the code in existing
scripts, so as to keep tracking easier (versioner_hashes).

Subject changed scripts to shellcheck.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2320685/) on ef25fe2 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
Forgot to update packages_config.yml

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2320844/) on 4c2dca4 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

Commit:
prevent libcxx from updating beyond 17.x

We need to prevent libcxx from updating to 18.x. Seems to cause
issues on Apple Si hardware.

Commit:
bump versioner hashes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2324056/) on e494ba8 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

roystgnr commented:
I'm just going to kick that Valgrind recipe.  Only one memory error, and it happens to be triggered by WebServerControl at a time when we've been having network issues?  That's got to be a false positive.

lindsayad commented:
Is this ready?!

milljm commented:
> Is this ready?!

Yes, lets merge this while its green!

milljm commented:
The instructions are not complete, but I can do that in another PR...

https://mooseframework.inl.gov/docs/PRs/26848/site/getting_started/installation/conda.html

This needs to say:

```bash
conda create -n moose moose-dev=2024.07.10 mpich
```
Otherwise, a person may pull the OpenMPI variant packages (and we are not testing those yet)

lindsayad commented:
Yea I really want this and the PETSc update to get in ASAP so that we can get moving on some NEAMS M2 needs
